### PR TITLE
added unit tests for sc-reconcile - cover getFscBySc method 

### PR DESCRIPTION
**Pull request by alonfirestein**

### CODE REVIEW

**Comment by alonfirestein** [comment]
[create unit-test for storageClass-reconcile - cover getFlashSystemClusterByStorageClass method](https://trello.com/c/4BDRWQCk/101-create-unit-test-for-storageclass-reconcile-cover-getflashsystemclusterbystorageclass-method)

**Comment by alonfirestein** [comment]
image

File: controllers/storageclass/storageclass_controller_test.go [hunk 1]
```diff
@@ -216,6 +216,118 @@ var _ = Describe("StorageClassWatcher", func() {
...
+
+			By("creating a new secret with topology awareness")
+			topologySecretName := "topology-secret"
+			secret := &corev1.Secret{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      topologySecretName,
+					Namespace: namespace,
+				},
+				Data: map[string][]byte{
+					"config": []byte("IHsKICAgImRlbW8tbWFuYWdlbWVudC1pZC0xIjogewogICAgICJ1c2VybmFtZSI6ICJkZW1vLXVzZX" +
```
Comment (tyichye19) [inline controllers/storageclass/storageclass_controller_test.go:277]: why encoded data?

File: controllers/storageclass/storageclass_controller_test.go [hunk 2]
```diff
@@ -216,6 +216,118 @@ var _ = Describe("StorageClassWatcher", func() {
 			}, timeout, interval).Should(BeTrue())
 		})
 
+		It("should get FlashSystemCluster by StorageClass successfully", func() {
+			ctx := context.TODO()
+
+			By("By querying the created StorageClass")
```
Comment (tyichye19) [inline controllers/storageclass/storageclass_controller_test.go:222]: why do we need to query the storageclass here?

File: controllers/storageclass/storageclass_controller_test.go [hunk 3]
```diff
@@ -216,6 +216,118 @@ var _ = Describe("StorageClassWatcher", func() {
...
+						"mRlbW8tdXNlcm5hbWUtMiIsCiAgICAgInBhc3N3b3JkIjogImRlbW8tcGFzc3dvcmQtMiIsCiAgICAgIm1hbmFnZW1l" +
+						"bnRfYWRkcmVzcyI6ICJkZW1vLW1hbmFnZW1lbnQtYWRkcmVzcy0yIiwKICAgICAic3VwcG9ydGVkX3RvcG9sb2dpZXM" +
+						"iOiBbCiAgICAgICB7CiAgICAgICAgICJ0b3BvbG9neS5ibG9jay5jc2kuaWJtLmNvbS9kZW1vLXJlZ2lvbiI6ICJkZW" +
+						"1vLXJlZ2lvbi0yIiwKICAgICAgICAgInRvcG9sb2d5LmJsb2NrLmNzaS5pYm0uY29tL2RlbW8tem9uZSI6ICJkZW1vL" +
+						"XpvbmUtMiIKICAgICAgIH0KICAgICBdCiAgIH0KIH0="),
+				},
+			}
+			Expect(k8sClient.Create(ctx, secret)).Should(Succeed())
+
+			By("By creating a new topology StorageClass")
```
Comment (tyichye19) [inline controllers/storageclass/storageclass_controller_test.go:292]: Not sure what we are trying to test here.
The idea is that we need to assume the objects were created successfully, and now we want to verify all other parts related to us are performed correctly (like adding the StorageClass to all relevant StorageSystem entries with the correct pool name, etc.)

File: controllers/storageclass/storageclass_controller_test.go [hunk 4]
```diff
@@ -216,6 +216,118 @@ var _ = Describe("StorageClassWatcher", func() {
...
+				err := k8sClient.Get(ctx, fscLookupKey, createdFsc)
+				return err == nil
+			}, timeout, interval).Should(BeTrue())
+
+			Expect(createdFsc.Name).To(Equal(FlashSystemName))
+			Expect(createdFsc.Namespace).To(Equal(namespace))
+			Expect(createdFsc.Spec.Secret.Name).To(Equal(secretName))
+
+			By("creating a new secret with topology awareness")
+			topologySecretName := "topology-secret"
```
Comment (tyichye19) [inline controllers/storageclass/storageclass_controller_test.go:270]: please use const for `topologySecretName`

File: controllers/storageclass/storageclass_controller_test.go [hunk 5]
```diff
@@ -216,6 +221,86 @@ var _ = Describe("StorageClassWatcher", func() {
 			}, timeout, interval).Should(BeTrue())
 		})
 
+		It("should get FlashSystemCluster by StorageClass successfully", func() {
```
Comment (tyichye19) [inline controllers/storageclass/storageclass_controller_test.go:224]: please add more test cases - like creating StorageClass with two matches FlashSystemClusters and different pool name.

File: controllers/storageclass/storageclass_controller_test.go [hunk 6]
```diff
@@ -216,6 +221,86 @@ var _ = Describe("StorageClassWatcher", func() {
...
+			Expect(k8sClient.Delete(ctx, createdTopologySc)).Should(Succeed())
+
+			By("By querying the ConfigMap to verify the topology StorageClass is deleted")
+			Eventually(func() bool {
+				err := k8sClient.Get(ctx, cmLookupKey, createdCm)
+				if err == nil {
+					var sp util.FlashSystemClusterMapContent
+					err = json.Unmarshal([]byte(createdCm.Data[FlashSystemName]), &sp)
+					if err == nil {
+						return sp.ScPoolMap[topologyStorageClassName] == ""
```
Comment (tyichye19) [inline controllers/storageclass/storageclass_controller_test.go:296]: I think the verification here should be like we did for regular StorageClass (check if key exist)

**Review by bvered — APPROVED** [review]

**Pull request merged**